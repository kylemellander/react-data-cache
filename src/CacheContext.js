import React, { createContext, useContext, useState } from 'react'

function CreateCache() {
  return new CacheImpl()
}

function CacheImpl() {
  this.resources = new Map()
  this.resolvers = new Map()
}

export const CacheContext = createContext()

export function CacheProvider({ children }) {
  // We are only using cache here to make sure that it persists through rerenders
  // It's possible we might do more in the future, but we cannot update the state and have it work
  // with our resource.
  const cacheState = useState(CreateCache())
  return (
    <CacheContext.Provider value={cacheState}>{children}</CacheContext.Provider>
  )
}

export function useForceRefresh() {
  const setCache = useContext(CacheContext)[1]
  return () =>
    setCache((cache) => {
      const newCache = CreateCache()
      newCache.resources = cache.resources
      newCache.resolvers = cache.resolvers
      return newCache
    })
}
