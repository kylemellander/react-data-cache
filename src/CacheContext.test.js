import { useContext } from 'react'
import { renderHook } from '@testing-library/react-hooks'
import '@testing-library/jest-dom/extend-expect'
import { CacheContext, CacheProvider } from './CacheContext'

describe('CacheProvider', () => {
  it('creates a cache with resources and resolvers', () => {
    const { result } = renderHook(() => useContext(CacheContext), {
      wrapper: CacheProvider
    })

    const cache = result.current[0]
    expect(cache.resources instanceof Map).toEqual(true)
    expect(cache.resolvers instanceof Map).toEqual(true)
    expect(cache).toMatchSnapshot()
  })
})
