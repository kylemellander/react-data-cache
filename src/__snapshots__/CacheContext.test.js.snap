// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CacheProvider creates a cache with resources and resolvers 1`] = `
CacheImpl {
  "resolvers": Map {},
  "resources": Map {},
}
`;
